app/models/file_depot_ftp.rb

```diff
@@ -125,7 +126,7 @@ def destination_path
 
   def base_path
     # uri: "ftp://ftp.example.com/incoming" => #<Pathname:incoming>
-    path = URI(URI.encode(uri)).path
+    path = URI(Addressable::URI.escape(uri)).path
```

Comment:
I'm curious why this was encoded in the first place...

```
[25] pry(main)> URI("ftp://ftp.example.com/incoming").path
=> "incoming"
[26] pry(main)> URI(URI.encode("ftp://ftp.example.com/incoming")).path
=> "incoming"
```

Reply:
Also note:

```ruby
[28] pry(main)> URI(ERB::Util.url_encode("ftp://ftp.example.com/incoming")).path
=> "ftp%3A%2F%2Fftp.example.com%2Fincoming"
[29] pry(main)> URI(CGI.escape("ftp://ftp.example.com/incoming")).path
=> "ftp%3A%2F%2Fftp.example.com%2Fincoming"
```

Reply:
Note that if this had had a fragment attached, it seems Addressable::URI is _better_ than URI.encode:

```ruby
[30] pry(main)> URI(URI.encode("ftp://ftp.example.com/incoming#fragment")).path
=> "incoming%23fragment"
[31] pry(main)> URI(Addressable::URI.encode("ftp://ftp.example.com/incoming#fragment")).path
=> "incoming"
```

Reply:
Well, I wrote that... so let me look into it...

Reply:
> Note that if this had had a fragment attached...

@Fryguy I don't think that a fragment would ever be valid for an FTP URL though.  But perhaps I am wrong.

Reply:
> Just kidding, no I didn't. It was all LJ 😄

Haha, now I have to see why I did it...

Reply:
@NickLaMuro found the original reason for the URI.encode:

```
commit 27492b325bb3892bb5b5bc1184d92c6e3c08f374
Author: Gregg Tanzillo <gtanzillo@manageiq.com>
Date:   Fri Sep 4 18:52:16 2009 +0000

    BugzID: 6494
    Added URI encode/decode wherever we handle log depot URIs and paths to address errors when posting to a URI that had spaces in the path.

    git-svn-id: http://miq-ubuntusub.manageiq.com/svn/svnrepos/Manageiq/trunk@16098 3c68ef56-dcc3-11dc-9475-a42b84ecc76f
```

Looking at the old FogBugz ticket I see these log lines:

```
[----] I, [2009-08-26T16:19:15.285493 #7703]  INFO -- : MIQ(LogFile-file_exists_ftp?) Checking existance of log file [ftp://ftp.manageiq.com/thenn/from_xav/20090826/Chicago_2/EVM ChiMgmt_1/Current_Chicago_2_EVM ChiMgmt_1_20090826_090208_20090826_161910.zip]...
[----] E, [2009-08-26T16:19:15.285798 #7703] ERROR -- : MIQ(LogFile-add_log_file_ftp) 'bad URI(is not URI?): ftp://ftp.manageiq.com/thenn/from_xav/20090826/Chicago_2/EVM ChiMgmt_1/Current_Chicago_2_EVM ChiMgmt_1_20090826_090208_20090826_161910.zip', writing to FTP: [ftp://ftp.manageiq.com/thenn/from_xav/20090826], Username: [thennessy@manageiq]
[----] I, [2009-08-26T16:19:15.295576 #7703]  INFO -- : MIQ(MiqServer-post_current_logs) Task: [1] Posting of current logs failed for EVM ChiMgmt Chicago MiqServer 1 due to error: [RuntimeError] ['bad URI(is not URI?): ftp://ftp.manageiq.com/thenn/from_xav/20090826/Chicago_2/EVM ChiMgmt_1/Current_Chicago_2_EVM ChiMgmt_1_20090826_090208_20090826_161910.zip', writing to FTP: [ftp://ftp.manageiq.com/thenn/from_xav/20090826], Username: [thennessy@manageiq]]
```

So it seems this was only added to handle spaces in incoming URLs.  If we add (or have) specs for spaces, then we can choose any encoder.

Reply:
Thanks for digging into this.  I was going to look at the various places we used it to find what specific characters were we concerned with.  I'll look at the remaining ones to see if it's the same story.

Reply:
Ok, I added tests demonstrating what worked previously with 2.6/2.7 with both URI.decode and Addressable::URI.decode.
After going through all of the commits, the only common thread was to deal with converting spaces to `%20` as @Fryguy pointed out.  I did the other escaped characters just for "fun".